tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,211 @@
+   */
+  public static boolean hasAutoScope() {
+    Iterator<TensorScope> iterator = scopeStack.get().descendingIterator();
+    while (iterator.hasNext()) {
+      TensorScope scope = iterator.next();
+      if (scope.autoAttach) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Detaches the given tensor from any scopes managing it, requiring it to be manually closed.
+   */
+  public static void detach(Tensor t) {
+    RawTensor raw = t.asRawTensor();
+    synchronized (TensorScope.class) {
+      allScopes.forEach(x -> x.detachTensor(raw));
+    }
+    raw.attached = false;
+  }
+
+  /**
+   * Create a new tensor scope.  If {@code autoAttach} is false, will not automatically manage tensors.
+   *
+   * @see TensorScope
+   */
+  public TensorScope(boolean autoAttach) {
+    this.autoAttach = autoAttach;
+
+    synchronized (TensorScope.class) {
+      allScopes.add(this);
+    }
+
+    localScopeStack = scopeStack.get();
+    localScopeStack.push(this);
+  }
+
+  /**
+   * Create a new tensor scope that automatically manages tensors.
+   */
+  public TensorScope() {
+    this(true);
+  }
+
+  /**
+   * Attach a tensor to this scope.  This happens automatically to tensors that are created in the scope.
+   * @return this
+   */
+  public TensorScope attach(Tensor t) {
+    RawTensor rt = t.asRawTensor();
+    rt.attached = true;
+    tensors.add(rt);
+
+    return this;
+  }
+
+  /**
+   * Attach tensors to this scope.  This happens automatically to tensors that are created in the scope.
+   * @return this
+   */
+  public TensorScope attach(Tensor... tensors) {
+    if (tensors != null) {
+      for (Tensor t : tensors) {
+        attach(t);
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * Attach tensors to this scope.  This happens automatically to tensors that are created in the scope.
+   * @return this
+   */
+  public TensorScope attach(HasTensors tensors) {
+    tensors.tensors().forEach(this::attach);
+
+    return this;
+  }
+
+  /**
+   * Attach tensors to this scope.  This happens automatically to tensors that are created in the scope.
+   * @return this
+   */
+  public TensorScope attach(HasTensors... tensors) {
+    if (tensors != null) {
+      for (HasTensors ht : tensors) {
+        attach(ht);
+      }
+    }
+
+    return this;
+  }
+
+  private void detachTensor(Tensor t) {
+    tensors.remove(t.asRawTensor());
+  }
+
+  private void closeScope() {
+    tensors.forEach(Tensor::close);
+
+    synchronized (TensorScope.class) {
+      allScopes.remove(this);
+    }
+
+    closed = true;
+  }
+
+  /**
+   * Closes this scope and its tensors, and any inner scopes.
+   */
+  @Override
+  public void close() {
+    if (closed) {
+      return;
+    }
+
+    if (!localScopeStack.contains(this)) {
+      throw new IllegalStateException("This scope is not on the scope stack, but was not closed."
+          + "  This should not be possible.");
```

Comment:
Why are you allowing scopes only on the stack??

Reply:
Since it closed child scopes by closing any scopes in front of it in the queue.  If somehow it got removed from the queue, that would silently close all scopes.  This shouldn't happen, but did before I started storing the stack as a local variable (if it was closed in a different thread than it was opened in), so this is just for safety.

Reply:
No, I mean, if `localScopeStack` doesn't contain `this`, then this test fails. That makes it impossible to use a given instance as a field and/or from some other thread! Doesn't it?

Reply:
Kind of.  `localScopeStack` is a local variable, set from the thread local when the scope is created.  It's not thread local itself, so `this` should always be in it.

It does mean that semantically, the scope is closed on the same thread it was opened on.  This makes some sense, because it's only going to collect tensors from that thread anyways.